kubespawner/spawner.py

```diff
@@ -1880,32 +1941,55 @@ def _start(self):
         return (pod["status"]["podIP"], self.port)
 
     @gen.coroutine
-    def stop(self, now=False):
-        delete_options = client.V1DeleteOptions()
-
-        if now:
-            grace_seconds = 0
-        else:
-            grace_seconds = self.delete_grace_period
+    def _make_delete_pod_request(self, pod_name, delete_options, grace_seconds, request_timeout):
+        """
+        Make an HTTP request to delete the given pod
 
-        delete_options.grace_period_seconds = grace_seconds
-        self.log.info("Deleting pod %s", self.pod_name)
+        Designed to be used with exponential_backoff, so returns
+        True / False on success / failure
+        """
+        self.log.info("Deleting pod %s", pod_name)
         try:
-            yield self.asynchronize(
+            yield gen.with_timeout(timedelta(seconds=request_timeout), self.asynchronize(
                 self.api.delete_namespaced_pod,
-                name=self.pod_name,
+                name=pod_name,
                 namespace=self.namespace,
                 body=delete_options,
                 grace_period_seconds=grace_seconds,
-            )
+            ))
+            return True
+        except gen.TimeoutError:
+            return False
         except ApiException as e:
             if e.status == 404:
                 self.log.warning(
                     "No pod %s to delete. Assuming already deleted.",
-                    self.pod_name,
+                    pod_name,
                 )
+                # If there isn't already a pod, that's ok too!
+                return True
             else:
                 raise
+
+    @gen.coroutine
+    def stop(self, now=False):
+        delete_options = client.V1DeleteOptions()
+
+        if now:
+            grace_seconds = 0
+        else:
+            grace_seconds = self.delete_grace_period
+
+        delete_options.grace_period_seconds = grace_seconds
+
+
+        yield exponential_backoff(
+            partial(self._make_delete_pod_request, self.pod_name, delete_options, grace_seconds, self.k8s_api_request_timeout),
+            f'Could not delete pod {self.pod_name}',
+            # FIXME: We should instead add a timeout_times property to exponential_backoff instead
```

Comment:
FWIW, I don't agree that this `timeout_times` logic belongs in exponential_backoff. exponential_backoff is well defined for retrying something up to a time limit, letting unhandled errors raise. Handling a separate, shorter timeout as one of the cases that should be retried rightly belongs in the internal function without any knowledge of that in the outer exponential_backoff logic.

Reply:
Yeah, I think setting an overall number_of_tries is the way to go.